app/models/settings/base.rb

```diff
@@ -0,0 +1,159 @@
+        key = key.to_s
+
+        raise(ProcetedKeyError, "Can't use #{key} as setting key.") if key.in?(PROTECTED_KEYS)
+
+        @defined_fields ||= []
+        @defined_fields << {
+          key: key,
+          default: default,
+          type: type || :string
+        }
+
+        # Getter
+        define_singleton_method(key) do
+          result = __send__(:value_of, key)
+          result ||= default.is_a?(Proc) ? default.call : default
+          result = __send__(:convert_string_to_value_type, type, result, separator: separator)
+
+          result
+        end
+
+        # Setter
+        define_singleton_method("#{key}=") do |value|
+          var_name = key
+
+          record = find_by(var: var_name) || new(var: var_name)
+          value = __send__(:convert_string_to_value_type, type, value, separator: separator)
+
+          record.value = value
+          record.save!
+
+          value
+        end
+
+        # Validation
+        if validates
+          validates[:if] = proc { |item| item.var.to_s == key }
+          __send__(:validates, key, **validates)
+
+          define_method(:read_attribute_for_validation) { |_key| self.value }
+        end
+
+        return unless type == :boolean
+
+        # Predicate method for booleans
+        define_singleton_method("#{key}?") { __send__(key) }
+      end
+
+      def convert_string_to_value_type(type, value, separator: nil)
+        return value unless value.class.in?([String, Integer, Float, BigDecimal])
+
+        case type
+        when :boolean
+          value.in?(["true", "1", 1, true])
+        when :array
+          value.split(separator || SEPARATOR_REGEXP).reject(&:empty?).map(&:strip)
+        when :hash
+          value = begin
+            YAML.safe_load(value).to_h
+          rescue StandardError
+            {}
+          end
+          value.deep_stringify_keys!
+          ActiveSupport::HashWithIndifferentAccess.new(value)
+        when :integer
+          value.to_i
+        when :float
+          value.to_f
+        when :big_decimal
+          value.to_d
+        else
+          value
+        end
+      end
+
+      def value_of(var_name)
+        unless Database.table_available?(table_name)
+          # Fallback to default value if table was not ready (before migrate)
+          Rails.logger.warn("'#{table_name}' does not exist, '#{name}.#{var_name}' will return the default value.")
+          return
+        end
+
+        all_settings[var_name]
+      end
+
+      def all_settings
+        RequestStore[cache_key] ||= Rails.cache.fetch(cache_key, expires_in: 1.week) do
+          unscoped.select(:var, :value).each_with_object({}) do |record, result|
+            result[record.var] = record.value
+          end.with_indifferent_access
+        end
+      end
+    end
+
+    # get the value field, YAML decoded
+    def value
+      YAML.load(self[:value]) if self[:value].present? # rubocop:disable Security/YAMLLoad
+    end
```

Comment:
what happened to https://github.com/huacnlee/rails-settings-cached/blob/7695c7f39b6616b7aa6c171425595e749da9955b/lib/rails-settings/base.rb#L139 and `YAML.safe_load` ?

Reply:
Re `_convert_string_to_typeof_value`: I renamed it to `convert_string_to_value_type` which IMO is a clearer name. That method still uses `YAML.safe_load`:

```ruby
        when :hash
          value = begin
            YAML.safe_load(value).to_h
          rescue StandardError
            {}
          end
          value.deep_stringify_keys!
          ActiveSupport::HashWithIndifferentAccess.new(value)
```

What you're commenting on here is the `value` instance methods. In the version we pinned these were at the bottom of the file, in the current version of the gem they are at the top. I kept them at the bottom because generally Ruby code has instance methods after class methods. As you can see `#value` still uses `YAML.load` in `rails-settings-cached` too:

https://github.com/huacnlee/rails-settings-cached/blob/7695c7f39b6616b7aa6c171425595e749da9955b/lib/rails-settings/base.rb#L18

I experimented with switching this over to `safe_load` but things started breaking. At least the original author got rid of `eval` which was still used in previous versions of the gem.

Reply:
Got it!